airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/file/ProtobufDataChannelReader.kt

```diff
@@ -4,33 +4,56 @@
 
 package io.airbyte.cdk.load.file
 
-import com.google.common.io.CountingInputStream
+import com.google.protobuf.CodedInputStream
 import io.airbyte.cdk.load.message.DestinationMessage
 import io.airbyte.cdk.load.message.DestinationMessageFactory
 import io.airbyte.protocol.protobuf.AirbyteMessage.AirbyteMessageProtobuf
 import java.io.InputStream
+import kotlin.NoSuchElementException
+
+class ProtobufDataChannelReader(
+    private val factory: DestinationMessageFactory,
+    private val bufferSize: Int = 16 * 1024,
+) : DataChannelReader {
 
-/**
- * Parses a stream of size-prefixed protobuf messages, yielding a sequence of sized
- * [DestinationMessage]'s.
- */
-class ProtobufDataChannelReader(private val destinationMessageFactory: DestinationMessageFactory) :
-    DataChannelReader {
     private val parser = AirbyteMessageProtobuf.parser()
-    override fun read(inputStream: InputStream): Sequence<DestinationMessage> = sequence {
-        val countingInputStream = CountingInputStream(inputStream)
-        var count = countingInputStream.count
-        while (true) {
-            val protoMessage = parser.parseDelimitedFrom(countingInputStream) ?: break
-            val newCount = countingInputStream.count
-            val serializedSizeBytes = newCount - count
-            count = newCount
-            yield(
-                destinationMessageFactory.fromAirbyteProtobufMessage(
-                    protoMessage,
-                    serializedSizeBytes
-                )
-            )
+
+    override fun read(inputStream: InputStream): Sequence<DestinationMessage> =
+        object : Sequence<DestinationMessage> {
+            override fun iterator(): Iterator<DestinationMessage> =
+                object : Iterator<DestinationMessage> {
+
+                    private val cis: CodedInputStream =
+                        CodedInputStream.newInstance(inputStream, bufferSize).apply {
+                            enableAliasing(false)
+                            setSizeLimit(Int.MAX_VALUE)
+                        }
+
+                    private var nextMsg: DestinationMessage? = fetch()
+
+                    override fun hasNext(): Boolean = nextMsg != null
+
+                    override fun next(): DestinationMessage {
+                        val out = nextMsg ?: throw NoSuchElementException()
+                        nextMsg = fetch()
+                        return out
+                    }
+
+                    private fun fetch(): DestinationMessage? {
+                        if (cis.isAtEnd) return null
```

Comment:
nit:  the return could be lifted out of the "if" and converted to an if/else to make more Kotlin-y/make the method = the if statement instead of having { }.